sql/core/src/main/scala/org/apache/spark/sql/streaming/ui/StreamingQueryStatisticsPage.scala

```diff
@@ -126,6 +126,54 @@ private[ui] class StreamingQueryStatisticsPage(parent: StreamingQueryTab)
     <br />
   }
 
+  def generateWatermark(
+      query: StreamingQueryUIData,
+      minBatchTime: Long,
+      maxBatchTime: Long,
+      jsCollector: JsCollector): NodeBuffer = {
+    // This is made sure on caller side but put it here to be defensive
+    require(query.lastProgress != null)
+    if (query.lastProgress.eventTime.containsKey("watermark")) {
+      val watermarkData = query.recentProgress.flatMap { p =>
+        val batchTimestamp = parseProgressTimestamp(p.timestamp)
+        val watermarkValue = parseProgressTimestamp(p.eventTime.get("watermark"))
+        if (watermarkValue > 0L) {
+          // seconds
+          Some((batchTimestamp, ((batchTimestamp - watermarkValue) / 1000.0)))
+        } else {
+          None
+        }
+      }
+      val maxWatermark = watermarkData.maxBy(_._2)._2
+      val graphUIDataForWatermark =
+        new GraphUIData(
+          "watermark-gap-timeline",
+          "watermark-gap-histogram",
+          watermarkData,
+          minBatchTime,
+          maxBatchTime,
+          0,
+          maxWatermark,
+          "seconds")
+      graphUIDataForWatermark.generateDataJs(jsCollector)
+
+      // scalastyle:off
+      new NodeBuffer() &+
+        <tr>
+          <td style="vertical-align: middle;">
+            <div style="width: 160px;">
+              <div><strong>Global Watermark Gap {SparkUIUtils.tooltip("The gap between timestamp and global watermark for the batch.", "right")}</strong></div>
```

Comment:
I understand that `timestamp` here means now but maybe we can more explicit.

Reply:
Yes. Probably better to say `batch timestamp` explicitly.